File: common/src/main/java/juuxel/adorn/criterion/SitOnBlockCriterion.java
package juuxel.adorn.criterion;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import net.minecraft.advancements.criterion.SimpleCriterionTrigger;
import net.minecraft.advancements.criterion.BlockPredicate;
import net.minecraft.advancements.criterion.EntityPredicate;
import net.minecraft.advancements.criterion.ContextAwarePredicate;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.core.BlockPos;

import java.util.Optional;

public final class SitOnBlockCriterion extends SimpleCriterionTrigger<SitOnBlockCriterion.Conditions> {
    @Override
    public Codec<Conditions> codec() {
        return Conditions.CODEC;
    }

    public void trigger(ServerPlayer player, BlockPos pos) {
        trigger(player, conditions -> conditions.matches(player, pos));
    }

    public record Conditions(Optional<ContextAwarePredicate> player, BlockPredicate block) implements SimpleCriterionTrigger.SimpleInstance {
        public static final Codec<Conditions> CODEC = RecordCodecBuilder.create(instance -> instance.group(
            EntityPredicate.ADVANCEMENT_CODEC.optionalFieldOf("player")
                .forGetter(Conditions::player),
            BlockPredicate.CODEC.fieldOf("block").forGetter(Conditions::block)
        ).apply(instance, Conditions::new));

        public boolean matches(ServerPlayer player, BlockPos pos) {
            return block.matches(player.level(), pos);
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/criterion/SitOnBlockCriterion.kt
package juuxel.adorn.criterion

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.advancement.criterion.AbstractCriterion
import net.minecraft.predicate.BlockPredicate
import net.minecraft.predicate.entity.EntityPredicate
import net.minecraft.predicate.entity.LootContextPredicate
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.dynamic.Codecs
import net.minecraft.util.math.BlockPos
import java.util.Optional

class SitOnBlockCriterion : AbstractCriterion<SitOnBlockCriterion.Conditions>() {
    override fun getConditionsCodec(): Codec<Conditions> = Conditions.CODEC

    fun trigger(player: ServerPlayerEntity, pos: BlockPos) {
        trigger(player) { it.matches(player, pos) }
    }

    data class Conditions(val playerPredicate: Optional<LootContextPredicate>, val block: BlockPredicate) : AbstractCriterion.Conditions {
        override fun player(): Optional<LootContextPredicate> = playerPredicate

        fun matches(player: ServerPlayerEntity, pos: BlockPos): Boolean =
            block.test(player.serverWorld, pos)

        companion object {
            val CODEC: Codec<Conditions> = RecordCodecBuilder.create { instance ->
                instance.group(
                    Codecs.createStrictOptionalFieldCodec(EntityPredicate.LOOT_CONTEXT_PREDICATE_CODEC, "player")
                        .forGetter(Conditions::playerPredicate),
                    BlockPredicate.CODEC.fieldOf("block").forGetter(Conditions::block)
                ).apply(instance, ::Conditions)
            }
        }
    }
}
